index/index.go

```diff
@@ -43,16 +43,97 @@ func New(path string) *Index {
 	return &index
 }
 
-// BuildIndex reads each of the files and adds a key -> (file, offset) pair
-// to the master index for every key in every file.
-func (index *Index) BuildIndex() error {
+// Load reads each of the files and adds a key -> (file, offset) pair to the
+// master index for every key in every file. If a manifest file and index
+// already exist in the directory, it'll use that.
+func (index *Index) Load() error {
 	err := index.buildFileList()
 	if err != nil {
 		return err
 	}
 
+	// Try loading and checking the manifest first.
+	manifestPath := filepath.Join(index.Path, ".manifest")
+	manifest, err := readManifest(manifestPath)
+	if err == nil {
+		log.Println("Loading index from existing manifest at", manifestPath)
+		err = index.loadIndexFromManifest(manifest)
+		if err != nil {
+			log.Println("Failed to load existing manifest with error:", err)
+		} else {
+			index.Ready = true
+			return nil
+		}
+	}
+
+	err = index.buildNewIndex()
+	if err != nil {
+		return err
+	}
+
+	index.Ready = true
+	return nil
+}
+
+func (index *Index) buildFileList() error {
+	infos, err := ioutil.ReadDir(index.Path)
+	if err != nil {
+		return err
+	}
+
+	index.files = make([]indexFile, 0, len(infos))
+	index.readLocks = make([]sync.Mutex, len(infos))
+
+	for _, info := range infos {
+		if !info.IsDir() && !strings.HasPrefix(info.Name(), "_") && !strings.HasPrefix(info.Name(), ".") {
+			err := index.addFile(info.Name())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (index *Index) loadIndexFromManifest(m manifest) error {
+	for i, manifestFile := range m.Files {
+		indexFile := index.files[i]
+		baseName := filepath.Base(indexFile.file.Name())
+		if baseName != filepath.Base(manifestFile.Name) {
+			return fmt.Errorf("unmatched file: %s", manifestFile.Name)
+		}
+
+		info, err := indexFile.file.Stat()
+		if err != nil {
+			return err
+		}
+
+		if info.Size() != manifestFile.Size {
```

Comment:
These are good questions. I think a hash would be better than mtime, since this is actually designed to be okay with you moving the data directory wholesale. CRC would probably be pretty quick to calculate, I bet.